src/Features/Core/Portable/DesignerAttribute/DesignerAttributeDiscoveryService.cs

```diff
@@ -79,57 +90,58 @@ public DesignerAttributeDiscoveryService()
+            AsyncLazy<VersionStamp> lazyProjectVersion,
+            AsyncLazy<bool> lazyHasDesignerCategoryType,
             CancellationToken cancellationToken)
         {
-            // We need to reanalyze the project whenever it (or any of its dependencies) have
-            // changed.  We need to know about dependencies since if a downstream project adds the
-            // DesignerCategory attribute to a class, that can affect us when we examine the classes
-            // in this project.
-            var projectVersion = await project.GetDependentSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
-
             // Now get all the values that actually changed and notify VS about them. We don't need
             // to tell it about the ones that didn't change since that will have no effect on the
             // user experience.
             var changedData = await ComputeChangedDataAsync(
-                project, specificDocument, projectVersion, designerCategoryType, cancellationToken).ConfigureAwait(false);
+                project, specificDocument, lazyProjectVersion, lazyHasDesignerCategoryType, cancellationToken).ConfigureAwait(false);
 
             // Only bother reporting non-empty information to save an unnecessary RPC.
             if (!changedData.IsEmpty)
-                await callback.ReportDesignerAttributeDataAsync(changedData, cancellationToken).ConfigureAwait(false);
+                await callback.ReportDesignerAttributeDataAsync(changedData.SelectAsArray(d => d.data), cancellationToken).ConfigureAwait(false);
 
             // Now, keep track of what we've reported to the host so we won't report unchanged files in the future. We
             // do this after the report has gone through as we want to make sure that if it cancels for any reason we
             // don't hold onto values that may not have made it all the way to the project system.
-            foreach (var data in changedData)
+            foreach (var (data, projectVersion) in changedData)
                 _documentToLastReportedInformation[data.DocumentId] = (data.Category, projectVersion);
         }
 
-        private async Task<ImmutableArray<DesignerAttributeData>> ComputeChangedDataAsync(
+        private async Task<ImmutableArray<(DesignerAttributeData data, VersionStamp version)>> ComputeChangedDataAsync(
             Project project,
             Document? specificDocument,
-            VersionStamp projectVersion,
-            INamedTypeSymbol designerCategoryType,
+            AsyncLazy<VersionStamp> lazyProjectVersion,
+            AsyncLazy<bool> lazyHasDesignerCategoryType,
             CancellationToken cancellationToken)
         {
-            using var _1 = ArrayBuilder<Task<DesignerAttributeData?>>.GetInstance(out var tasks);
+            using var _ = ArrayBuilder<(DesignerAttributeData data, VersionStamp version)>.GetInstance(out var results);
             foreach (var document in project.Documents)
```

Comment:
Probably add a comment here why we prefer to process documents sequentially instead of concurrently?

Reply:
sure.  will do.